### ISSUE TITLE
Molecule Property prediction task

### ISSUE THREAD

**Issue by salman-moh** [MolecularAI/Chemformer#13]
Since we dont have access to the weights of the fine-tuned model for molecule property prediction task, I ask how do we fine-tune for this task using the Chemformer codebase?

I see for finetune we use:  `python -m molbart.example_scripts.finetune_regression.finetuneRegr.py <args>`

Below are just the default args, but I'd like to know how would the data_path file look like for molecule property prediction task?

> DEFAULT_data_path = 'all_TrainTestVal_pXC50.csv'
> DEFAULT_vocab_path = "prop_bart_vocab.txt"
> DEFAULT_study_name = "TL_Regr_pXC50_" + str(datetime.datetime.now())
> DEFAULT_model_path = "pt_zinc_bart.ckpt"
> DEFAULT_genes_path = "/individual_files"
> DEFAULT_BATCH_SIZE = 32
> DEFAULT_ACC_BATCHES = 1
> DEFAULT_GRAD_CLIP = 1.0
> DEFAULT_SCHEDULE = "cycle"
> DEFAULT_AUGMENT = True
> DEFAULT_WARM_UP_STEPS = 3000
> DEFAULT_TRAIN_TOKENS = None
> DEFAULT_NUM_BUCKETS = 24
> DEFAULT_LIMIT_VAL_BATCHES = 1.0
> DEFAULT_EPOCHS = 150 
> DEFAULT_GPUS = 1
> DEFAULT_D_PREMODEL = 512
> DEFAULT_MAX_SEQ_LEN = 300
> DEFAULT_LR = 3e-4
> DEFAULT_H_FEEDFORWARD = 2048
> DEFAULT_drp = 0.2 
> DEFAULT_Hdrp = 0.4 
> DEFAULT_WEIGHT_DECAY = 0.0

I have the 3 datasets with me: FreeSolv, Lipophilicity and ESOL.

![image](https://user-images.githubusercontent.com/31631107/197244149-ddb5a812-5df8-4ca1-ac25-fd79eb32400b.png)
![image](https://user-images.githubusercontent.com/31631107/197244223-5f3722d9-8608-4359-948c-61527bfd2674.png)
![image](https://user-images.githubusercontent.com/31631107/197244298-b66e558b-0043-4188-a1fc-f0201b6cf3cf.png)

Thanks in advance !

**Comment by EBjerrum** [comment]
@spirosdim

**Comment by spirosdim** [comment]
Hi @salman-moh,

Your `data_path` file should have the following columns: [SMILES, pXC50, SET]. 
For example (these SMILES are from the ExCAPE database):
![df](https://user-images.githubusercontent.com/31864574/197385796-8e90d0d1-067c-427f-ae5e-3753db1c760a.jpg)


The SMILES column is created by concatenating the name of the property/gene-activity and the SMILES, e.g. `<FreeS>|C1CCCCC1`. Notes: we used the pipe symbol `|` to be easily separated for the augmentation part later in the dataloader, the target `pXC50` is scaled as described in the paper.

You might find it helpful to read section 2.3 in the [paper](https://iopscience.iop.org/article/10.1088/2632-2153/ac3ffb).
Also, you can find the special tokens in the vocabulary [prop_bart_vocab.txt](https://github.com/MolecularAI/Chemformer/blob/main/example_scripts/finetune_regression/prop_bart_vocab.txt), you can add your own special tokens by replacing the unused ones and then fine-tune. 
Finally, for more details check the [RegPropDataset](https://github.com/MolecularAI/Chemformer/blob/main/example_scripts/finetune_regression/finetune_regression_modules.py#L32) and [RegPropDataModule](https://github.com/MolecularAI/Chemformer/blob/main/example_scripts/finetune_regression/finetune_regression_modules.py#L73).

**Comment by salman-moh** [comment]
I'm confused, this is for _biological activity tasks_ or _molecular property prediction_?
I thought this was for the  _biological activity tasks_ with the pXC50 values, and my question was about the _molecular property prediction_.
Apologies if my interpretation is wrong.

**Comment by spirosdim** [comment]
You can use the same code and adapt it for the property prediction tasks. 
The column name `pXC50` denotes the target, we just kept it as is from the biological activity tasks. It should be interpreted as the target.

**Comment by salman-moh** [comment]
As described in the paper, I'm trying to get all SMILES which are in more than 1 dataset of FreeSolv/Lipo/ESOL and I only found 120 SMILES out of which 10 from FreeSol, 50 from ESOL and 60 from Lipo.. so after upsample it would still be only 30 FreeSol,100 ESOL and 60 Lipo datapoints. Am I doing something wrong?

Using `Chem.CanonSmiles()` function from RDKit to find a match between the 3 datasets `Chem.CanonSmiles(lipo_row['SMILES']) == Chem.CanonSmiles(frees_row[' SMILES'])` `Chem.CanonSmiles(lipo_row['SMILES']) == Chem.CanonSmiles(esol_row[' SMILES'])`  to check if compound in Lipo exists in ESOL/FreeSolv datasets.

**Comment by spirosdim** [comment]
Perhaps it is not stated clearly, I will try to rephrase it. There is no need to get only the SMILES that appear in more than 1 dataset of FreeSolv/Lipo/ESOL. You should use all the observations from the datasets you want to fine-tune.

The paper states: "we ensure that all molecules which appear in more than one dataset appear only in the train set" which means that the train set has the molecules which appear in more than one dataset and other molecules that appear in only one dataset, but the validation and test sets do not have molecules which appear in more than one dataset.

Additionally, "Due to the size imbalance between the three datasets, the ESOL and Free Solvation datasets were upsampled by factors of 2 and 3, respectively, during training." which means, after creating the train/validation/test splits, only in the train set the observations from ESOL and Free Solvation datasets were upsampled by factors of 2 and 3 respectively.

**Comment by spirosdim** [comment]
1. First all SMILES were turned into their canonical representation, then their strings were compared (with value_counts of all SMILES and then by taking the ones that appear more than one time).

2. After scaling, it does not matter if some examples from the validation/test set have values slightly greater than 1 or slightly less than 0. The point is that we should scale each task separately to be in the same range more or less (0 to 1). Additionally, generally scaling the values helps the model in training.

**Comment by spirosdim** [comment]
You successfully fine-tuned your model. 
The error here corresponds to the calculation and tracking of metrics in each _biological activity task_.
You need to modify the function [results_133QSAR](https://github.com/MolecularAI/Chemformer/blob/main/example_scripts/finetune_regression/finetuneRegr.py#L136) to your tasks.

**Comment by salman-moh** [comment]
Yeah, I have modified it but my files had some mistakes, works now.
![image](https://user-images.githubusercontent.com/31631107/198358011-41084499-a790-4e57-88c9-478ec9273353.png)
I wonder why R2 is missing. (trained for 150 epochs, the initial model was combined pretrained model)

**Comment by salman-moh** [comment]
ok maybe it was something wrong with the dataset? the fine-tuned model shows the R2 values now.. 
![image](https://user-images.githubusercontent.com/31631107/199084629-3632d7af-a0c3-4209-bb8b-d9ad8babf4e2.png)

But @spirosdim could you please confirm if the [pred.py](https://github.com/salman-moh/Chemformer/blob/e247bb75834626ae0540a463ed1d5c0ff9d704b7/example_scripts/finetune_regression/pred.py) is the correct way to do inference? Its a bit lengthy with the csv file and the dataloader, because the model expects a dataloader, so I'm wondering for a single string inference how do we go about doing the inference? For example: `input="<ESOL>|Clc1ccc(C(c2ccc(Cl)cc2)C(Cl)Cl)cc1"`

**Comment by spirosdim** [comment]
The [pred.py](https://github.com/salman-moh/Chemformer/blob/e247bb75834626ae0540a463ed1d5c0ff9d704b7/example_scripts/finetune_regression/pred.py) is one way to do batch inference. 

You can modify the [RegPropDataset](https://github.com/salman-moh/Chemformer/blob/e247bb75834626ae0540a463ed1d5c0ff9d704b7/example_scripts/finetune_regression/finetune_regression_modules.py#L32), [RegPropDataModule](https://github.com/salman-moh/Chemformer/blob/e247bb75834626ae0540a463ed1d5c0ff9d704b7/example_scripts/finetune_regression/finetune_regression_modules.py#L73) and [get_targs_preds](https://github.com/salman-moh/Chemformer/blob/e247bb75834626ae0540a463ed1d5c0ff9d704b7/example_scripts/finetune_regression/finetuneRegr.py#L110) to receive a single string for inference.

**Comment by salman-moh** [comment]
Thanks so much for the help to both!

**Comment by salman-moh** [comment]
@spirosdim had to re-open the issue due to some problem with prediction, I was trying to convert [pred.py](https://github.com/salman-moh/Chemformer/blob/e247bb75834626ae0540a463ed1d5c0ff9d704b7/example_scripts/finetune_regression/pred.py) into real-time inference by modifying `RegPropDataset`, but I noticed:
For a given example for every run, produces different results
![image](https://user-images.githubusercontent.com/31631107/201397785-e40849b4-6e71-47db-bd89-462fdb9d876e.png)
![image](https://user-images.githubusercontent.com/31631107/201397858-9340157d-deee-4a87-8cc0-16d28e9c5288.png)
is there something wrong I am doing? this happens even with batch inference using the [pred.py](https://github.com/salman-moh/Chemformer/blob/e247bb75834626ae0540a463ed1d5c0ff9d704b7/example_scripts/finetune_regression/pred.py)

**Comment by spirosdim** [comment]
hi @salman-moh 
it might be because the `pred.py` uses `augment=True` in `RegPropDataModule`
for not using augmentation on input SMILES try `augment=False`  [here](https://github.com/salman-moh/Chemformer/blob/e247bb75834626ae0540a463ed1d5c0ff9d704b7/example_scripts/finetune_regression/pred.py#L55)

**Comment by salman-moh** [comment]
I tried that, same issue.
![image](https://user-images.githubusercontent.com/31631107/201508791-a1594dcc-a0f7-415e-9e4c-2c77b127810b.png)
![image](https://user-images.githubusercontent.com/31631107/201508804-98956f64-667b-4d00-9ad2-e9fda8459c5b.png)
I dont think this is supposed to happen right? I cant recall any stochastic component in the model..

**Comment by spirosdim** [comment]
it might be the dropout or the layer normalization
try to add after [this line](https://github.com/salman-moh/Chemformer/blob/e247bb75834626ae0540a463ed1d5c0ff9d704b7/example_scripts/finetune_regression/finetuneRegr.py#L119) the model in evaluation mode (turns of dropout and any batch/layer normalization)  and use no gradients (this is just for computational reasons) with the following:

```
model.eval()
with torch.no_grad():
	for i, batch in enumerate(iter(dl)):
        batch['encoder_input'] = batch['encoder_input'].to(device) 
		...
```

**Comment by salman-moh** [comment]
ah yes, this works, thank you :)

**Comment by salman-moh** [comment]
Re-opening because I'm re-finetuning with the above merged repo, and since the team was curious about the performance, will post the metrics soon.

**Comment by salman-moh** [comment]
Not sure why, but after finishing running the `finetuneRegr.py` script, the predictions are constant... with the [get_targs_preds() ](https://github.com/salman-moh/Chemformer/blob/e247bb75834626ae0540a463ed1d5c0ff9d704b7/example_scripts/finetune_regression/finetuneRegr.py#L110)function. 
I am printing only 5 datapoints for each dataset, ESOL, LIPO, FREES. Both training and test parts.
![image](https://user-images.githubusercontent.com/31631107/214109987-e89ee1be-5b54-4771-b02d-c57e034e2c48.png)

Not sure if @spirosdim you encountered this while fine-tuning? But I cant tell where this issue is coming from, perhaps something wrong in `finetune_regression_modules.py` ?

**Comment by spirosdim** [comment]
I did not encounter this problem before. This reminds me of a [previous comment ](https://github.com/MolecularAI/Chemformer/issues/13#issuecomment-1293846535) on this issue. 

You could check if there is any problem with the input data format.

**Comment by SGenheden** [comment]
Closing due to inactivity
